File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque


def solution(places):
    answer = []
    
    # 상 하 좌 우
    mx = [0, 0, -1, 1]
    my = [1, -1, 0, 0]
    
    # 두 교육생 정보 입력받아 거리 체크하고, 맨하탄 거리 2 이하면 파티션 체크하기
    def check_partition(sx, sy, dx, dy, room):
        visited = [[False for _ in range(len(room))] for __ in range(len(room))]
        visited[sx][sy] = True            
        dq = deque()
        dq.append([sx, sy, False, 0])
            
        while dq:
            cx, cy, pfound, moved = dq.popleft()
            
            
            if cx == dx and cy == dy and moved == 2:
                if not pfound:
                    #print(f"not pfound, cx : {cx}, cy : {cy}")
                    return False
                    break
            
    
            for m in range(4):
                visited[dy][dx] = False
                nx, ny = cx + mx[m], cy + my[m]
                if 0 <= nx < len(room) and 0 <= ny < len(room) and not visited[ny][nx]:
                    visited[ny][nx] = True
                    if room[ny][nx] == 'X' or pfound:
                        dq.append([nx, ny, True, moved+1])
                    else:
                        dq.append([nx, ny, False, moved+1])
        
        return True
    
    
    # 고사장을 받아서 정보 체크하기
    def check_distanced(room):
        people = []
        for i in range(len(room)):
            for j in range(len(room)):
                if room[j][i] == 'P':
                    people.append([i, j])
        # print(people)
        
        for i in range(0, len(people)-1):
            for j in range(i+1, len(people)):
                x1, y1, x2, y2 = people[i][0], people[i][1], people[j][0], people[j][1]
                dist = abs(x2 - x1) + abs(y2 - y1)
                if dist < 2:
                    return False
                elif dist == 2:
                    if x1 == x2 and room[min(y1, y2)+1][x1] != 'X':
                        return False
                    elif y1 == y2 and room[y1][min(x1, x2)+1] != 'X':
                        return False
                    else:                        
                        if not check_partition(x1, y1, x2, y2, room):
                            return False
        #print("distanced checked, True")
        return True
        
    
    for place in places:
        if check_distanced(place):
            answer.append(1)
        else:
            answer.append(0)
    return answer

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    public int[] solution(String[][] places) {
        int[] answer = new int[places.length];
        
        for(int i = 0; i < places.length; i++){
            String[] place = places[i];
            answer[i] = checkDistancing(place) ? 1 : 0;
        }
        
        return answer;
    }
    
    public boolean checkDistancing(String[] place) {
        ArrayList<int[]> persons = new ArrayList<>();
        for(int i = 0; i < place.length; i++){
            for(int j = 0; j < place[i].length(); j++){
                if(place[i].charAt(j) == 'P') persons.add(new int[] {i, j});
            }
        }
        
        for(int i = 0; i < persons.size() - 1; i++){
            for(int j = i + 1; j < persons.size(); j++){
                int p1r = persons.get(i)[0];
                int p1c = persons.get(i)[1];
                int p2r = persons.get(j)[0];
                int p2c = persons.get(j)[1];            
                
                if(getMenDist(p1r, p1c, p2r, p2c) < 2) return false;
                else if (getMenDist(p1r, p1c, p2r, p2c) > 2) continue;
                else {
                    if(!checkPartition(p1r, p1c, p2r, p2c, place)) return false;
                }
            }
        }
        
        return true;
    }
    
    public int getMenDist(int r1, int c1, int r2, int c2){
        return Math.abs(r1 - r2) + Math.abs(c1 - c2);
    }

    public boolean checkPartition(int p1r, int p1c, int p2r, int p2c, String[] place){
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        
        ArrayDeque<int[]> deque = new ArrayDeque<>();
        boolean[][] visited = new boolean[5][5];
        deque.add(new int[] {p1r, p1c});
        visited[p1r][p1c] = true;
        
        while(!deque.isEmpty()){
            int cr = deque.peek()[0];
            int cc = deque.peek()[1];
            deque.poll();
            
            if(cr == p2r && cc == p2c) return false;
            
            for(int d = 0; d < 4; d++){
                int nr = cr + dr[d];
                int nc = cc + dc[d];
                
                if(nr < 0 || 5 <= nr || nc < 0 || 5 <= nc || nr < Math.min(p1r, p2r) || Math.max(p1r, p2r) < nr 
                   || nc < Math.min(p1c, p2c) || Math.max(p1c, p2c) < nc  || visited[nr][nc] || place[nr].charAt(nc) == 'X') continue;                
                
                visited[nr][nc] = true;
                deque.add(new int[] {nr, nc});
            }
        }
        
        return true;
    }
}